README.md
Twig
========================================

Opinionated minimal logging inspired by and compatible with [node-bunyan](https://github.com/trentm/node-bunyan)

----------------------------------------

##  Features

 - JSON formatted, [bunyan](https://github.com/trentm/node-bunyan)-conformant log entries
 - asynchronous logging, powered by [Kovenant](http://kovenant.komponents.nl/) and [Disruptor](https://lmax-exchange.github.io/disruptor/)
 - console only by default
 - easily extendable and customisable via simple Appender interface
 - startup configuration via optional environment variables


### Basic

Having a TestObject
```kotlin
object TestObject {
    val logger: Logger = Logger(this)

    fun dummyFunction() {
        logger.info("dummy")
    }
}
```
and calling its ```dummyFunction``` will result in a log entry similar to
```json
{"hostname":"vyo-pc","pid":6484,"thread":"Thread[main,5,]","time":"2015-11-30T18:55:35.092Z","level":30,"name":"io.github.vyo.twig.TestObject@12aba81","msg":"dummy","v":0}
```

### Custom Fields

You may pass in an arbitrary number of ```Pair<String, Any>``` to create additional custom log entry fields

```
logger.info("dummy", Pair("my custom field", "my custom content"), Pair("my logger", logger))
```

```
{"hostname":"vyo-pc","pid":6156,"thread":"Thread[main,5,]","time":"2015-11-30T19:09:58.507Z","level":30,"name":"io.github.vyo.twig.TestObject@16e898f","msg":"dummy","my custom field":"my custom content","my logger":"io.github.vyo.twig.logger.Logger@1a97992","v":0}
```

### Auto-expanded Throwables

You may pass throwables, i.e. exceptions and errors, and they will be automatically expanded to both
a short message and an accompanying stacktrace.

Stacktraces will only be logged out at DEBUG level (configurable) or below. They will be put into a custom field
called '*stacktrace*' in the form of an array. 

```
val logger = Logger("twig")

try {
    Integer.parseInt("not an int")
} catch (e: Exception) {
    logger.error(e)
    logger.debug(e)
}
```
will lead to
```
{"hostname":"vyo-pc"","pid":6580,"thread":"Thread[main,5,]","time":"2015-11-30T19:38:45.037Z","level":50,"name":"twig","msg":"java.lang.NumberFormatException: For input string: \"not an int\"","v":0}
{"hostname":"vyo-pc"","pid":6580,"thread":"Thread[main,5,]","time":"2015-11-30T19:38:45.037Z","level":20,"name":"twig","msg":"java.lang.NumberFormatException: For input string: \"not an int\"","stacktrace":["java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)","java.lang.Integer.parseInt(Integer.java:580)","java.lang.Integer.parseInt(Integer.java:615)"],"v":0}
```

### Bunyan CLI

Piped into the bunyan cli the preceding log entry will be pretty-printed like this

```
[2015-11-30T19:09:58.507Z]  INFO: io.github.vyo.twig.TestObject@16e898f/6156 on vyo-pc: dummy (thread=Thread[main,5,], my custom field="my custom content", my logger=io.github.vyo.twig.logger.Logger@1a97992)
```

### Configuration

__*Note*__: You may configure global options by referencing either ```Logger.*``` or ```Logger.global.*```.

The configuration is logged on startup; by default 
 - the global log level will be INFO, 
 - the worker amount will be set to the system's available processors,
 - the log queue size will be set to 1024,
 - throwables will be auto-expanded at DEBUG level or below,
 - and auto-expanded throwables will come with a stacktrace of 50 lines at most.

```
{"hostname":"vyo-pc","pid":1144,"thread":"Thread[main,5,]","time":"2015-11-30T19:36:28.220Z","level":30,"name":"twig","msg":"logging worker count: 4","v":0}
{"hostname":"vyo-pc","pid":1144,"thread":"Thread[main,5,]","time":"2015-11-30T19:36:28.220Z","level":30,"name":"twig","msg":"logging work queue size: 1024","v":0}
{"hostname":"vyo-pc","pid":1144,"thread":"Thread[main,5,]","time":"2015-11-30T19:36:28.220Z","level":30,"name":"twig","msg":"global log level: INFO","v":0}
{"hostname":"vyo-pc"","pid":6580,"thread":"Thread[main,5,]","time":"2015-11-30T19:38:45.037Z","level":30,"name":"twig","msg":"throwable expansion level: DEBUG","v":0}
{"hostname":"vyo-pc"","pid":6580,"thread":"Thread[main,5,]","time":"2015-11-30T19:38:45.037Z","level":30,"name":"twig","msg":"throwable expansion depth: 50","v":0}
```

Setting up the following environment variables in advance
```sh
TWIG_LEVEL=TRACE
TWIG_QUEUE=64
TWIG_WORKERS=1
TWIG_EXPANSION_LEVEL=TRACE
TWIG_EXPANSION_DEPTH=100
```

```
{"hostname":"vyo-pc"","pid":6580,"thread":"Thread[main,5,]","time":"2015-11-30T19:38:45.037Z","level":30,"name":"twig","msg":"global log level TRACE","v":0}
{"hostname":"vyo-pc"","pid":6580,"thread":"Thread[main,5,]","time":"2015-11-30T19:38:45.037Z","level":30,"name":"twig","msg":"logging work queue size: 64","v":0}
{"hostname":"vyo-pc"","pid":6580,"thread":"Thread[main,5,]","time":"2015-11-30T19:38:45.037Z","level":30,"name":"twig","msg":"logging worker count: 1","v":0}
{"hostname":"vyo-pc"","pid":6580,"thread":"Thread[main,5,]","time":"2015-11-30T19:38:45.037Z","level":30,"name":"twig","msg":"throwable expansion level: TRACE","v":0}
{"hostname":"vyo-pc"","pid":6580,"thread":"Thread[main,5,]","time":"2015-11-30T19:38:45.037Z","level":30,"name":"twig","msg":"throwable expansion depth: 100","v":0}
```

Re-assigning the global log level or appender will also be logged
```
Logger.global.appender = ConsoleAppender()
Logger.global.level = Level.WARN
```

```
{"hostname":"vyo-pc"","pid":6364,"thread":"Thread[main,5,]","time":"2015-11-30T19:40:36.876Z","level":30,"name":"twig","msg":"global appender io.github.vyo.twig.appender.ConsoleAppender@1ae6ba4","v":0}
{"hostname":"vyo-pc"","pid":6364,"thread":"Thread[main,5,]","time":"2015-11-30T19:40:36.876Z","level":30,"name":"twig","msg":"global log level WARN","v":0}
```

You may also specify your own serialiser function if the built in JSON-serialiser does not fit your needs.

####Default (built in):
```
Logger.global.serialiser = Logger.global.simpleSerialiser
```
####Jodd:
```
val jodd = JsonSerializer()
Logger.global.serialiser = { any: Any -> jodd.serialize(any) }
```
####Jackson:
```
val jackson = ObjectMapper()
Logger.global.serialiser = { any: Any -> jackson.writeValueAsString(any) }
```
####Custom function:
```
fun customToJson( any: Any) : String {
    return any.toString()
}
Logger.global.serialiser = { any: Any -> customToJson(any) }

// or:

val lambdaJson = { any: Any -> any.toString() }
Logger.global.serialiser = { any: Any -> lambdaJson(any) }

// or:
Logger.global.serialiser = { any: Any -> any.toString() }
```
__*Note*__: You should ensure that your custom function produces valid JSON.
You may also want to have arrays and collections be automatically expanded,
especially when making use of **Twig**'s auto-expanded throwables.

__*Note*__: Google's **GSON** seems to be unable to handle cyclic references, making it unsuitable for **Twig** for the time being.


### Per-logger settings:

Log level, appender and serialiser can be set globally, as well as individually for each logger:
```
val loggerOne = Logger("one")
val loggerTwo = Logger("two")

loggerOne.level = Level.WARN
loggerTwo.level = Level.INFO

loggerOne.appender = FileAppender()
loggerTwo.appender = ConsoleAppender()

loggerOne.serialiser = Logger.simpleSerialiser
loggerTwo.serialiser = { any: Any -> any.toString() }

```

### Exception Behaviour

If an exception occurs during the actual logging process, e.g. because the underlying Appender fails, we try to log a diagnostic entry to STDERR
```
logger.fatal("exceptional")
```
```
{"hostname":"vyo-pc","pid":3092,"thread":"Thread[main,5,]","time":"2015-11-30T19:47:15.128Z","level":60"name":"io.github.vyo.twig.logger.Logger@50ea2a","msg":"logging failed: null","original level":30,"original name":"io.github.vyo.twig.TestObject@970b10","original message":"exceptional","v":0
```

##  Notes

This project adheres to the [semantic versioning](http://semver.org/) and [change log](http://keepachangelog.com/) guidelines.

##  Author

Manuel Weidmann (@vyo)

##  License

ISC License


CHANGELOG.md
#   change log

This project adheres to the [semantic versioning](http://semver.org/) and [change log](http://keepachangelog.com/) guidelines.

##  [Unreleased]
### Added
  - custom message fields may now be restricted by log level

### Fixed
  - stacktrace expansion checked against global log level instead of log level of the message

##  [v1.1.0] - 2016-07-16
### Added
- allow specifying serialisation function
- add option to auto-expand throwables
- properly handle arrays and collections

##  [v1.0.6] - 2016-07-12
### Changed
- switch to kotlin 1.0.0 and kovenant 3.0.0

##  [v1.0.5] - 2015-12-07
### Fixed
- log level threshold was effectively ignored

##  [v1.0.4] - 2015-12-02
### Changed
- made kovenant dependency versions explicit and variable

##  [v1.0.3] - 2015-12-02
### Changed
- re-generated gradle wrapper scripts
### Fixed
- superfluous was compiled into the jar

##  [v1.0.2] - 2015-12-02
### Added
- exception log entries extended to contain all original entry fields
### Fixed
- special characters in string were not properly escaped for JSON

##  [v1.0.1] - 2015-11-30
### Fixed
- logger creation from Java required passing in an Appender

##  [v1.0.0] - 2015-11-30
### Added
- startup configuration via environment variables
    -   TWIG_LEVEL: initial global log level; default is ```INFO```
    -   TWIG_QUEUE: work queue size, will be rounded up to a power of two; default is ```1024```
    -   TWIG_WORKERS: maximum number of concurrent tasks; default is ```number of processors```
### Changed
- Logger.root is now Logger.global

##  [v0.11.0] - 2015-11-30
### Changed
- drop superfluous LoggerInterface
- improve consistency by renaming ```threshold``` to ```level```
- Logger.root is not a logger itself anymore
- Logger.root internals are now hidden, excepting
    -   appender
    -   threshold

##  [v0.10.0] - 2015-11-29
### Added
- log calls now return ```Promise<Unit, Exception>``` instead of ```Unit```:
    -   log entry creation can now be waited on
    -   log failures can now be manually handled in addition to the default diagnostic logging to STDERR
### Changed
- error log entries will always be written to STDERR instead of the root logger's appender
### Fixed
- no log entries were created if no custom fields were passed in
- log error handling was disabled

##  [v0.9.0] - 2015-11-29
### Changed
- log message may now be of type Any
- log message may not be empty
- custom field usage changed:
    -   custom fields are not specified in advance anymore
    -   custom fields are now passed on as a variable number of Pair<String, Any> representing pairs of custom field name and custom field value, e.g. as ```log.info("my message", Pair("customField", "my custom content"))``` or ```log.info("product bought", Pair("price", 5.99))```

##  [v0.8.0] - 2015-11-29
### Added
- log entries are bunyan-parseable
- log entries have additional fields
    -   hostname
    -   pid
    -   v
### Changed
- loggers inherit the root logger's Appender by default, i.e. ConsoleAppender
- log entry elememts changed
    -   value of key 'time' was in milliseconds, is now ISO-formatted:  yyyy-MM-dd'T'HH:mm:ss.SSS'Z'
    -   key 'message' changed to 'msg'

##  [v0.7.0] - 2015-11-28
### Added
- allow using of custom Appenders

##  [v0.6.0] - 2015-11-28
### Added
- allow adding of custom fields on loggers
- log changes of root log level
### Changed
- package structure adjusted for upcoming Appender extension

##  [v0.5.2] - 2015-11-28
### Fixed
- formatting bug which caused log entries not to be valid JSON objects
### Added
- log root log level on Logger initialisation

##   [v0.5.1] - 2015-11-28
### Added
- minimal test specification via the [Spek](http://jetbrains.github.io/spek/) library

##   [v0.5.0] - 2015-11-28
### Fixed
- logger initialisation possible again (root logger initialisation competed with Logger instance initialisation)

### Added
- new interface LoggerInterface
### Changed
- Logger implements LoggerInterface
- root logger implements LoggerInterface instead of
   extending Logger
- root logger manually delegates calls to private logger

##   [v0.4.1] - 2015-11-28
### Added
- log configuration information on initialisation
- Disruptor support
    - queue size of 1024 (default)
    - worker thread counts equals number of available processors


##   [v0.3.0] - 2015-11-28
### Changed
- Logger.root object is now a Logger itself

##   [v0.2.0] - 2015-11-26
### Changed
- root log level no longer takes precedence

##   [v0.1.2] - 2015-11-26
### Changed
- newly created loggers now inherit the log level of the root logger

##  [v0.1.1] - 2015-11-26
### Changed
- default root log level change:
    - from DEBUG
    - to INFO

##   [v0.1.0] - 2015-11-25
### Added
- console logging
- JSON log entries
- global log level threshold
- per-logger log level threshold (root log level takes precedence)


src/main/kotlin/io/github/vyo/twig/logger/Logger.kt
package io.github.vyo.twig.logger

import io.github.vyo.twig.appender.Appender
import io.github.vyo.twig.appender.ConsoleAppender
import nl.komponents.kovenant.Kovenant
import nl.komponents.kovenant.Promise
import nl.komponents.kovenant.disruptor.queue.disruptorWorkQueue
import nl.komponents.kovenant.task
import java.io.PrintWriter
import java.io.StringWriter
import java.lang.management.ManagementFactory
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.*

/**
 * Default logger implementation for Twig
 *
 * Created by Manuel Weidmann on 24.11.2015.
 */

open class Logger @JvmOverloads constructor(val caller: Any,
                                            var appender: Appender = Logger.global.appender,
                                            var level: Level = Logger.global.level,
                                            var serialiser: (any: Any) -> String = Logger.global.simpleSerialiser) {

    companion object global {

        private val TWIG_LEVEL = "TWIG_LEVEL"
        private val TWIG_WORKERS = "TWIG_WORKERS"
        private val TWIG_QUEUE = "TWIG_QUEUE"
        private val TWIG_EXPANSION_LEVEL = "TWIG_EXPANSION_LEVEL"
        private val TWIG_EXPANSION_DEPTH = "TWIG_EXPANSION_DEPTH"

        val simpleSerialiser = { any: Any ->

            when (any) {
                is Boolean,
                is Double,
                is Float,
                is Long,
                is Int,
                is Short,
                is Byte,
                is Char -> simplePrimitiveSerialiser(any)
                is Array<*> -> simpleArraySerialiser(any)
                is Collection<*> -> simpleCollectionSerialiser(any)
                else -> "\"${escapeSpecialChars(any.toString())}\""
            }

        }

        private fun simplePrimitiveSerialiser(any: Any): String {
            return "${any.toString()}"
        }

        private fun simpleArraySerialiser(array: Array<*>): String {
            var string = "["

            for (element: Any? in array) {
                if (element is Any) {
                    string += simpleSerialiser(element)
                    string += ","
                }
            }

            string += "]"

            return string.replace(",]", "]")
        }

        private fun simpleCollectionSerialiser(collection: Collection<*>): String {
            var string = "["

            for (element: Any? in collection) {
                if (element is Any) {
                    string += simpleSerialiser(element)
                    string += ","
                }
            }

            string += "]"

            return string.replace(",]", "]")
        }

        var appender: Appender = ConsoleAppender()
            set(value) {
                field = value
                logger.appender = value
                logger.info("global appender $appender")
            }

        var level: Level = Level.INFO
            set(value) {
                field = value
                logger.info("global log level $level")
            }

        var serialiser: (any: Any) -> String = simpleSerialiser

        var expansionLevel = Level.DEBUG
            set(value) {
                field = value
                logger.info("global throwable expansion level $expansionLevel")
            }

        var expansionDepth = 50
            set(value) {
                field = value
                logger.info("global throwable expansion depth $expansionDepth")
            }

        private val processInfo: String = ManagementFactory.getRuntimeMXBean().name
        private val pid: Int = Integer.parseInt(processInfo.split('@')[0])
        private val hostName: String = processInfo.split('@')[1]
        private val timeZone: TimeZone = TimeZone.getTimeZone("UTC")
        private val isoFormat: DateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
        private val logger: Logger = Logger("twig")

        private fun escapeSpecialChars(string: String): String {

            var escapedString: String = string
            escapedString = escapedString.replace("\\", "\\\\")
            escapedString = escapedString.replace("\n", "\\n")
            escapedString = escapedString.replace("\r", "\\r")
            escapedString = escapedString.replace("\b", "\\b")
            escapedString = escapedString.replace("\t", "\\t")
            escapedString = escapedString.replace("\"", "\\\"")

            return escapedString
        }

        init {
            isoFormat.timeZone = timeZone

            val queue: Int
            val workers: Int

            val levelEnv: String? = System.getenv(TWIG_LEVEL)
            try {
                if (levelEnv is String) {
                    level = Level.valueOf(levelEnv)
                }
            } catch (exception: IllegalArgumentException) {
                level = Level.INFO
            }

            val queueEnv: String? = System.getenv(TWIG_QUEUE)
            if (queueEnv is String && Integer.parseInt(queueEnv) is Int) {
                queue = Integer.parseInt(queueEnv)
            } else {
                queue = 1024
            }

            val workerEnv: String? = System.getenv(TWIG_WORKERS)
            if (workerEnv is String && Integer.parseInt(workerEnv) is Int) {
                workers = Integer.parseInt(workerEnv)
            } else {
                workers = Runtime.getRuntime().availableProcessors()
            }

            Kovenant.context {
                callbackContext {
                    dispatcher {
                        concurrentTasks = workers
                        workQueue = disruptorWorkQueue(capacity = queue)
                    }
                }
            }

            val expansionLevelEnv: String? = System.getenv(TWIG_EXPANSION_LEVEL)
            try {
                if (expansionLevelEnv is String) {
                    expansionLevel = Level.valueOf(expansionLevelEnv)
                }
            } catch (exception: IllegalArgumentException) {
                level = Level.DEBUG
            }

            val expansionDepthEnv: String? = System.getenv(TWIG_EXPANSION_DEPTH)
            if (expansionDepthEnv is String && Integer.parseInt(expansionDepthEnv) is Int) {
                expansionDepth = Integer.parseInt(expansionDepthEnv)
            } else {
                expansionDepth = 50
            }

            logger.info("logging worker count: $workers")
            logger.info("logging work queue size: $queue")
            logger.info("global log level: $level")
            logger.info("throwable expansion level: $expansionLevel")
            logger.info("throwable expansion depth: $expansionDepth")
        }
    }

    private fun filterCustomMessagesByLevel(customMessages: Array<out Triple<String, Any, Level>>, level: Level):
            Array<Pair<String, Any>> {
        return customMessages.filter {
            level >= it.third
        }.map {
            Pair(it.first, it.second)
        }.toTypedArray()
    }

    fun log(level: Level, message: Any, vararg customMessages: Pair<String, Any>): Promise<Unit, Exception> {
        if (level < this.level) return task { }
        val thread: Thread = java.lang.Thread.currentThread()
        val time: String = isoFormat.format(Date(System.currentTimeMillis()))

        if (expansionLevel >= level && message is Throwable) {
            val stacktraceSize = Math.min(message.stackTrace.size, expansionDepth)

            val stacktrace = Arrays.copyOf(message.stackTrace, stacktraceSize)
            val adjustedCustomMessages = Arrays.copyOf(customMessages, customMessages.size + 1)

            adjustedCustomMessages[customMessages.size] = Pair("stacktrace", stacktrace)

            return log(level, "$message", *adjustedCustomMessages)
        }

        return task {
            var entry: String = "{${serialiser("hostname")}:${serialiser(hostName)}," +
                    "${serialiser("pid")}:${serialiser(pid)}," +
                    "${serialiser("thread")}:${serialiser(thread)}," +
                    "${serialiser("time")}:${serialiser(time)}," +
                    "${serialiser("level")}:${serialiser(level.toInt())}," +
                    "${serialiser("name")}:${serialiser(caller)}," +
                    "${serialiser("msg")}:${serialiser(message)}"

            for (customMessage in customMessages) {
                entry += ",${serialiser(customMessage.first)}:${serialiser(customMessage.second)}"
            }

            entry += ",${serialiser("v")}:${serialiser(0)}}"
            appender.write(entry)
        } fail {
            //get the stacktrace
            val writer: StringWriter = StringWriter()
            it.printStackTrace(PrintWriter(writer))
            val stacktrace = serialiser(writer.toString())

            var entry: String = "{${serialiser("hostname")}:${serialiser(hostName)}," +
                    "${serialiser("pid")}:${serialiser(pid)}," +
                    "${serialiser("thread")}:${serialiser(thread)}," +
                    "${serialiser("time")}:${serialiser(isoFormat.format(Date(System.currentTimeMillis())))}," +
                    "${serialiser("level")}:${serialiser(Level.FATAL.toInt())}" +
                    "${serialiser("name")}:${serialiser(this)}," +
                    "${serialiser("msg")}:${serialiser("logging exception: ${it.javaClass}")}," +
                    "${serialiser("exception message")}:${serialiser("${it.message}")}," +
                    "${serialiser("exception stacktrace")}:${serialiser(stacktrace)}," +
                    "${serialiser("original time")}:${serialiser(time)}," +
                    "${serialiser("original level")}:${serialiser(level.toInt())}," +
                    "${serialiser("original name")}:${serialiser(caller)}," +
                    "${serialiser("original msg")}:${serialiser(message)}"

            for (customMessage in customMessages) {
                entry += ",${serialiser("original " + customMessage.first)}:${serialiser(customMessage.second)}"
            }

            entry += ",${serialiser("v")}:${serialiser(0)}}"
            System.err.println(entry)
        }
    }


    fun trace(message: Any, vararg customMessages: Pair<String, Any>): Promise<Unit, Exception> {
        return log(Level.TRACE, message, *customMessages)
    }

    fun trace(message: Any, customMessages: Array<Triple<String, Any, Level>>): Promise<Unit, Exception> {
        return log(Level.TRACE, message, *filterCustomMessagesByLevel(customMessages, Level.TRACE))
    }

    fun debug(message: Any, vararg customMessages: Pair<String, Any>): Promise<Unit, Exception> {
        return log(Level.DEBUG, message, *customMessages)
    }

    fun debug(message: Any, customMessages: Array<Triple<String, Any, Level>>): Promise<Unit, Exception> {
        return log(Level.DEBUG, message, *filterCustomMessagesByLevel(customMessages, Level.DEBUG))
    }

    fun info(message: Any, vararg customMessages: Pair<String, Any>): Promise<Unit, Exception> {
        return log(Level.INFO, message, *customMessages)
    }

    fun info(message: Any, customMessages: Array<Triple<String, Any, Level>>): Promise<Unit, Exception> {
        return log(Level.INFO, message, *filterCustomMessagesByLevel(customMessages, Level.INFO))
    }

    fun warn(message: Any, vararg customMessages: Pair<String, Any>): Promise<Unit, Exception> {
        return log(Level.WARN, message, *customMessages)
    }

    fun warn(message: Any, customMessages: Array<Triple<String, Any, Level>>): Promise<Unit, Exception> {
        return log(Level.WARN, message, *filterCustomMessagesByLevel(customMessages, Level.WARN))
    }

    fun error(message: Any, vararg customMessages: Pair<String, Any>): Promise<Unit, Exception> {
        return log(Level.ERROR, message, *customMessages)
    }

    fun error(message: Any, customMessages: Array<Triple<String, Any, Level>>): Promise<Unit, Exception> {
        return log(Level.ERROR, message, *filterCustomMessagesByLevel(customMessages, Level.ERROR))
    }

    fun fatal(message: Any, vararg customMessages: Pair<String, Any>): Promise<Unit, Exception> {
        return log(Level.FATAL, message, *customMessages)
    }

    fun fatal(message: Any, customMessages: Array<Triple<String, Any, Level>>): Promise<Unit, Exception> {
        return log(Level.FATAL, message, *filterCustomMessagesByLevel(customMessages, Level.FATAL))
    }
}
